### DISCUSSION TITLE
MultiGraph should not extend Graph

### DISCUSSION THREAD

**Discussion by patrick-nicodemus** [networkx/networkx#6269] [Ideas]
I have the following constructive criticism: MultiGraph should not extend Graph. I believe that this is an LSP violation. The Liskov Substitution Principle may not be an ironclad law but I think it is a good practice which facilitates modular design and should be adhered to when possible.

Some of these violations are syntactic, for example add_edge and get_edge take a key parameter for a MultiGraph, but they do not take this parameter in a Graph, so there is a type signature mismatch. This reflects the underlying semantic difference between the function calls. if you have an existing edge from u -> v and you tried to perform `G.add_edge(u, v, key="a")` this would have a completely different meaning depending on whether G is a MultiGraph or a Graph - in one you have a new edge in addition to the old one keyed by `a`, in the other you have a single edge replacing the old one where the attribute dictionary is `a`. The differences between these behaviors are part of the publicly available interface because you can count how many edges there are in the graph.

The authors themselves point out that "NetworkX algorithms designed for weighted graphs cannot use multigraphs correctly because it is not clear how to handle multiedge weights." This is another illustration of how it violates LSP to have MultiGraph extend Graph.

Basically the inheritance function here is functioning as an easy form of code reuse

The use of the function `is_multigraph()` to control this seems ad-hoc to me. In my opinion, it would be better if Graph and MultiGraph were simply two separate classes, perhaps both extending some common class which incorporated the common functions of node management and declared abstract functions for both of them. Or there could be a NodeSet class, and both Graph and MultiGraph incorporate a NodeSet as a field of each instance, and also implement an edge management interface.

Thank you for your consideration.

**Comment by dschult** [comment]
I can assure you that an awful lot of time was spent on questions like this during the design and iterative building, followed by near complete rewriting, followed by iterating, etc. The classes do not follow LSP in the ways you specify, and perhaps some others. And it is true that we leaned more on using the classes as an easy form of code reuse/maintenance -- to the point that even the doc_strings are largely the same for the methods across the different classes. I don't recall many of the decisions made during those discussions. But I do recall that creating some sort of base class for the others to build off of slowed the code significantly. I suspect LSP was either created for compiled code rather than interpreted code, or that the design principles did not consider performance.  And to be fair, it may be that Python has evolved to the point where these performance issues are no longer significant -- though my occasional explorations into performance suggest otherwise. 

I'm sure we could improve the code style and type-handling and order of function arguments. But I'd much rather we spend our time making simpler interfaces, better documentation, more readable and performant code and implementing faster algorithms. So, I'm hoping that as you read through the codebase you make note of changes that could affect these aspects -- along with your notes about function signatures and class inheritance. 

Maybe I should also ask about this Issue -- with this posting are you looking for stories about how these decisions were made? Or suggesting a replacement? Or just saying how you would have done it? What is the goal of this Issue?

**Reply by patrick-nicodemus** [reply]
I made it a discussion rather than posting it as an issue because I didn't want to pin it to some specific resolution, "I will not be satisfied until xyz things are done." I did want some feedback from the project maintenance team because I was curious, and I thought if nothing else it would be helpful to have the issue documented here publicly together with your response. I was also wondering if people who depend on the library would pop in to say "Yes, this would be helpful for me" or "No this is irrelevant."

My interest in LSP was more in terms of benefits of reasoning about the code behavior with static analysis tools than Python performance, I do not really know the performance specs. I was just going through the files with mypy and thinking about how to handle some of the errors it threw, specifically that MultiGraph.add_edge has a different type signature than Graph.add_edge, this could be fixed in some superficial way perhaps but I realized that really they have fundamentally different meanings and so just rearranging the syntax of the input list would address the letter of the error but not the spirit. For my own applications I'll just mute the warnings that mypy throws for that stuff.

I realize that rearchitecting the code would be a substantial burden but I also didn't think it would be appropriate to say "that is too big to fix so I will not bother saying anything" so I wrote it up.
Thank you very much for your response.

**Reply by rossbar** [reply]
My personal 2 cents is that these design principles are of course important and should be considered when *designing* an API (and they certainly were, as noted above), but they are less likely to be actionable for such a mature library due to the potential upheaval in user code. In my personal opinion, this is *especially* pronounced in the case where there is not a substantial, tangible upside for the user. If the redesign would result in a 100x performance boost for users, then that redesign proposal would likely receive a lot of attention (assuming it conformed to the other core goals of the library). Retroactively conforming to design principles for the sake of static code analysis seems like a very poor tradeoff for the potential upheaval in user code (again, just my opinion).

FWIW the LSP issue has cropped up in other places (always in the context of static type checkers). It's worth noting that Python has added language features e.g. keyword-only arguments since the original design[^1] that would help with this issue and in many cases boost readability. It would be great to look into adopting this but again, this has substantial potential for upheaval of user code. Fortunately there's precedence and a blueprint for kwarg-only adoption in [scikit-learn](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep009/proposal.html). I'd be personally very interested in pursuing something along these lines, and I think the LSP stuff can serve as additional motivation, but this is a big change that would need strong championing through the NXEP proposal process.

Anyways, thanks for sharing feedback! Just wanted to add my drops to the bucket.

[^1]: The fact that keyword-only arguments can be considered a "new" language feature in the context of NetworkX development should give some idea of just how far back the development history spans :joy:
